core/Symbols.cc

```diff
@@ -1400,7 +1400,9 @@ vector<std::pair<NameRef, SymbolRef>> Symbol::membersStableOrderSlow(const Globa
     return result;
 }
 
-SymbolData::SymbolData(Symbol &ref, const GlobalState &gs) : DebugOnlyCheck(gs), symbol(ref) {}
+SymbolData::SymbolData(Symbol &ref, GlobalState &gs) : DebugOnlyCheck(gs), symbol(ref) {}
+
+ConstSymbolData::ConstSymbolData(const Symbol &ref, const GlobalState &gs) : DebugOnlyCheck(gs), symbol(ref) {}
```

Comment:
So this is the magic, right? It's a type error to construct a `SymbolData` if you don't have a `const GlobalState&`?

Reply:
The magic is that you must now have a non-const `GlobalState` to construct a `SymbolData`.

Previously, you could construct a `const SymbolData` with a `const GlobalState`. However, you can trivially copy a `const SymbolData` into a `SymbolData` and have mutable access to the symbol. It turns out that the compiler would do this transformation for you with the `auto` keyword, which is some bad stuff.

Reply:
So if we did `const auto symData` everywhere it would work?

Reply:
More-or-less. You could still accidentally copy it into a non-const `SymbolData` but we don't typically use `SymbolData` objects like that.